src/tallies/trigger.cpp

```diff
@@ -0,0 +1,203 @@
+#include "openmc/tallies/trigger.h"
+
+#include <cmath>
+#include <sstream>
+#include <utility> // for std::pair
+
+#include "openmc/capi.h"
+#include "openmc/constants.h"
+#include "openmc/error.h"
+#include "openmc/reaction.h"
+#include "openmc/settings.h"
+#include "openmc/simulation.h"
+#include "openmc/tallies/tally.h"
+
+namespace openmc {
+
+//==============================================================================
+// Global variable definitions
+//==============================================================================
+
+namespace settings {
+  KTrigger keff_trigger;
+}
+
+//==============================================================================
+// Non-member functions
+//==============================================================================
+
+std::pair<double, double>
+get_tally_uncertainty(int i_tally, int score_index, int filter_index)
+{
+  const auto& tally {model::tallies[i_tally]};
+
+  auto sum = tally->results_(filter_index, score_index, RESULT_SUM);
+  auto sum_sq = tally->results_(filter_index, score_index, RESULT_SUM_SQ);
+
+  int n = tally->n_realizations_;
+  auto mean = sum / n;
+  double std_dev = std::sqrt((sum_sq/n - mean*mean) / (n-1));
+  double rel_err = (mean != 0.) ? std_dev / std::abs(mean) : 0.;
+
+  return {std_dev, rel_err};
+}
+
+//! Find the limiting limiting tally trigger.
+//
+//! param[out] ratio The uncertainty/threshold ratio for the most limiting
+//!   tally trigger
+//! param[out] tally_id The ID number of the most limiting tally
+//! param[out] score The most limiting tally score bin
+
+void
+check_tally_triggers(double& ratio, int& tally_id, int& score)
+{
+  ratio = 0.;
+  for (auto i_tally = 0; i_tally < model::tallies.size(); ++i_tally) {
+    const Tally& t {*model::tallies[i_tally]};
+
+    // Ignore tallies with less than two realizations.
+    if (t.n_realizations_ < 2) continue;
+
+    for (const auto& trigger : t.triggers_) {
+      const auto& results = t.results_;
+      for (auto filter_index = 0; filter_index < results.shape()[0];
+           ++filter_index) {
+        for (auto score_index = 0; score_index < results.shape()[1];
+             ++score_index) {
+          // Compute the tally uncertainty metrics.
+          auto uncert_pair = get_tally_uncertainty(i_tally, score_index,
```

Comment:
We aren't using this if `trigger.metric == TriggerMetric::not_active`. We should probably re-factor the code to only calculate what we need.